src/AMDGPUOffload.cpp

```diff
@@ -0,0 +1,599 @@
+        std::vector<Expr> params;
+        params.push_back(module_state());
+        params.push_back(pipeline_name);
+        params.push_back(state_var_ptr(hex_name, type_of<int>()));
+        params.push_back(Call::make(type_of<size_t*>(), Call::make_struct, arg_sizes, Call::Intrinsic));
+        params.push_back(Call::make(type_of<void**>(), Call::make_struct, arg_ptrs, Call::Intrinsic));
+        params.push_back(Call::make(type_of<int*>(), Call::make_struct, arg_flags, Call::Intrinsic));
+
+        return call_extern_and_assert("halide_amdgpu_run", params);
+    }
+
+    Expr visit(const Let *op) override {
+        if (op->value.type() == Int(32)) {
+            alignment_info.push(op->name, modulus_remainder(op->value, alignment_info));
+        }
+
+        Expr expr = IRMutator2::visit(op);
+
+        if (op->value.type() == Int(32)) {
+            alignment_info.pop(op->name);
+        }
+        return expr;
+    }
+
+    Stmt visit(const LetStmt *op) override {
+        if (op->value.type() == Int(32)) {
+            alignment_info.push(op->name, modulus_remainder(op->value, alignment_info));
+        }
+
+        Stmt stmt = IRMutator2::visit(op);
+
+        if (op->value.type() == Int(32)) {
+            alignment_info.pop(op->name);
+        }
+        return stmt;
+    }
+
+public:
+    InjectAmdgpuRpc(Module &device_code) : device_code(device_code) {}
+
+    Stmt inject(Stmt s) {
+        s = mutate(s);
+
+        if (!device_code.functions().empty()) {
+            // Wrap the statement in calls to halide_initialize_kernels.
+            Expr runtime_buf_var = Variable::make(type_of<struct halide_buffer_t *>(), runtime_module_name + ".buffer");
+            Expr runtime_size = Call::make(Int(32), Call::buffer_get_extent, { runtime_buf_var, 0 }, Call::Extern);
+            Expr runtime_ptr = Call::make(Handle(), Call::buffer_get_host, { runtime_buf_var }, Call::Extern);
+
+            Expr code_buf_var = Variable::make(type_of<struct halide_buffer_t *>(), pipeline_module_name + ".buffer");
+            Expr code_size = Call::make(Int(32), Call::buffer_get_extent, { code_buf_var, 0 }, Call::Extern);
+            Expr code_ptr = Call::make(Handle(), Call::buffer_get_host, { code_buf_var }, Call::Extern);
+            Stmt init_kernels = call_extern_and_assert("halide_amdgpu_initialize_kernels",
+                                                       { module_state_ptr(), code_ptr, cast<uint64_t>(code_size), runtime_ptr, cast<uint64_t>(runtime_size) });
+            s = Block::make(init_kernels, s);
+        }
+
+        // TODO: This can probably go away due to general debug info at the submodule compile level.
+        debug(1) << "Amdgpu device code module: " << device_code << "\n";
+
+        return s;
+    }
+};
+
+}  // namespace
+
+Stmt inject_amdgpu_rpc(Stmt s, const Target &host_target,
+                        Module &containing_module) {
+    Target target(Target::Linux, Target::X86, 64);
```

Comment:
This should be the AMD GPU target OS and architecture. If you don't need any OS support (no thread pool, IO, etc.) then maybe you can just use Target::NoOS.

Reply:
After some debugging, the function `resolve_submodules()` is not calling `compile_to_buffer()` as there are no submodules. Should I do,
```cpp
Module Module::resolve_submodules() const {
if(target.has_feature(Target::AMDGPUGFX900)) {
  Module lowered_module(name(), target());
  lowered_module_append(this-> compile_to_amdgpu_shared_object(*this));
  return lowered_module;
}
if(submodules().empty()) {
  return *this;
}
.....
}
```

Reply:
@dsharletg can I add this code?